django/urls/resolvers.py

```diff
@@ -246,51 +246,61 @@ def __str__(self):
 
 
 _PATH_PARAMETER_COMPONENT_RE = _lazy_re_compile(
-    r"<(?:(?P<converter>[^>:]+):)?(?P<parameter>[^>]+)>"
+    r"""
+        <
+        # converter
+        (?:([^>:]+):)?
+        # parameter
+        ([^>]+)
+        >
+    """,
+    re.VERBOSE,
 )
 
+whitespace_set = frozenset(string.whitespace)
 
+
+@functools.lru_cache
 def _route_to_regex(route, is_endpoint=False):
     """
     Convert a path pattern into a regular expression. Return the regular
     expression and a dictionary mapping the capture names to the converters.
     For example, 'foo/<int:pk>' returns '^foo\\/(?P<pk>[0-9]+)'
     and {'pk': <django.urls.converters.IntConverter>}.
     """
-    original_route = route
     parts = ["^"]
+    all_converters = get_converters()
     converters = {}
-    while True:
-        match = _PATH_PARAMETER_COMPONENT_RE.search(route)
-        if not match:
-            parts.append(re.escape(route))
-            break
-        elif not set(match.group()).isdisjoint(string.whitespace):
+    previous_end = 0
+    for match in _PATH_PARAMETER_COMPONENT_RE.finditer(route):
+        if not whitespace_set.isdisjoint(match[0]):
             raise ImproperlyConfigured(
                 "URL route '%s' cannot contain whitespace in angle brackets "
-                "<…>." % original_route
+                "<…>." % route
             )
-        parts.append(re.escape(route[: match.start()]))
-        route = route[match.end() :]
-        parameter = match["parameter"]
+        raw_converter, parameter = match.groups()
         if not parameter.isidentifier():
             raise ImproperlyConfigured(
                 "URL route '%s' uses parameter name %r which isn't a valid "
-                "Python identifier." % (original_route, parameter)
+                "Python identifier." % (route, parameter)
             )
-        raw_converter = match["converter"]
         if raw_converter is None:
             # If a converter isn't specified, the default is `str`.
             raw_converter = "str"
         try:
-            converter = get_converter(raw_converter)
+            converter = all_converters[raw_converter]
         except KeyError as e:
             raise ImproperlyConfigured(
-                "URL route %r uses invalid converter %r."
-                % (original_route, raw_converter)
+                "URL route %r uses invalid converter %r." % (route, raw_converter)
```

Comment:
Since you're changing something on this line anyway...

```suggestion
                f"URL route {route!r} uses invalid converter {raw_converter!r}."
```